router/dataplane.go

```diff
@@ -752,66 +777,190 @@ func (d *DataPlane) Run(ctx context.Context) error {
 						continue
 					}
 					if result.OutConn == nil { // e.g. BFD case no message is forwarded
+						// @ unfold d.validResult(result, addrAliasesPkt)
+						// @ ghost if addrAliasesPkt {
+						// @ 	apply acc(result.OutAddr.Mem(), R15) --* acc(sl.AbsSlice_Bytes(tmpBuf, 0, len(tmpBuf)), R15)
+						// @ }
+						// @ sl.CombineRange_Bytes(p.Buffers[0], 0, p.N, writePerm)
+						// @ msgs[:pkts][i0].IsActive = false
+						// @ fold msgs[:pkts][i0].Mem()
 						continue
 					}
 
 					// Write to OutConn; drop the packet if this would block.
 					// Use WriteBatch because it's the only available function that
 					// supports MSG_DONTWAIT.
+					// @ unfold d.validResult(result, addrAliasesPkt)
+					// @ unfold writeMsgInv(writeMsgs)
 					writeMsgs[0].Buffers[0] = result.OutPkt
+					// @ writeMsgs[0].WildcardPerm = !addrAliasesPkt
+					// @ writeMsgs[0].IsActive = true
 					writeMsgs[0].Addr = nil
 					if result.OutAddr != nil { // don't assign directly to net.Addr, typed nil!
 						writeMsgs[0].Addr = result.OutAddr
 					}
-
+					// @ sl.NilAcc_Bytes()
+					// @ fold acc(writeMsgs[0].Mem(), R50)
+					// @ assert forall i int :: { &writeMsgs[i] } 0 <= i && i < len(writeMsgs) ==>
+					// @ 	acc(writeMsgs[i].Mem(), R50)
 					_, err = result.OutConn.WriteBatch(writeMsgs, syscall.MSG_DONTWAIT)
+					// @ unfold acc(writeMsgs[0].Mem(), R50)
+					// @ ghost if addrAliasesPkt && result.OutAddr != nil {
+					// @	apply acc(result.OutAddr.Mem(), R15) --* acc(sl.AbsSlice_Bytes(tmpBuf, 0, len(tmpBuf)), R15)
+					// @ }
+					// @ sl.CombineRange_Bytes(p.Buffers[0], 0, p.N, writePerm)
+					// @ msgs[:pkts][i0].IsActive = false
+					// @ fold msgs[:pkts][i0].Mem()
+					// @ fold writeMsgInv(writeMsgs)
 					if err != nil {
+						// @ requires err != nil && err.ErrorMem()
+						// @ decreases
+						// @ outline (
 						var errno /*@@@*/ syscall.Errno
-						if !errors.As(err, &errno) ||
+						// @ assert acc(&errno)
+						// @ fold errno.Mem()
+						errorsAs := errors.As(err, &errno)
+						// @ unfold errno.Mem()
+						if !errorsAs ||
 							!(errno == syscall.EAGAIN || errno == syscall.EWOULDBLOCK) {
 							log.Debug("Error writing packet", "err", err)
 							// error metric
 						}
+						// @ )
+						// @ assert acc(inputCounters.DroppedPacketsTotal.Mem(), _)
+						// @ prometheus.CounterMemImpliesNonNil(inputCounters.DroppedPacketsTotal)
 						inputCounters.DroppedPacketsTotal.Inc()
 						continue
 					}
+					// @ requires acc(&d, _)
+					// @ requires acc(MutexInvariant(d), _)
+					// @ requires result.EgressID in d.getDomForwardingMetrics()
+					// @ decreases
+					// @ outline(
 					// ok metric
+					// @ d.getForwardingMetricsMem(result.EgressID)
+					// @ unfold acc(forwardingMetricsMem(d.forwardingMetrics[result.EgressID], result.EgressID), _)
 					outputCounters := d.forwardingMetrics[result.EgressID]
+					// @ assert acc(outputCounters.OutputPacketsTotal.Mem(), _)
```

Comment:
Are these asserts that just check for a permission amount to a predicate necessary?

Reply:
They are not necessary, but I found them useful for debugging whenever I was converging to the specs. With this, I can find immediately whenever some expectation is broken